### PR TITLE
Feature/pmd and spotbugs gateway

### PR DESCRIPTION
**Pull request by simonhir**
**Description**

Add PMD and Spotbugs for refarch-gateway and fix warnings.

**References**
Issue #57, #58

### CODE REVIEW

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java [hunk 1]
```diff
@@ -30,51 +31,53 @@
...
                         .withBody("{ \"testkey\" : \"testvalue\" }")));
 
         webTestClient.get().uri("/api/refarch-gateway-backend-service/remote").exchange()
                 .expectStatus().isEqualTo(HttpStatus.OK)
-                .expectHeader().valueMatches("Content-Type", "application/json")
-                .expectHeader().doesNotExist("WWW-Authenticate")
-                .expectHeader().valueMatches("Expires", "0")
+                .expectHeader().valueMatches(org.springframework.http.HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType())
+                .expectHeader().doesNotExist(org.springframework.http.HttpHeaders.WWW_AUTHENTICATE)
+                .expectHeader().valueMatches(org.springframework.http.HttpHeaders.EXPIRES, "0")
```
Comment (devtobi) [inline refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java:80]: Please don't use the FQCN here and instead import correctly. :)

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/route/BackendRouteTest.java [hunk 2]
```diff
@@ -48,20 +50,21 @@ void setup() {
...
+                .header(XSRF_HEADER_NAME, "5cfb01a3-b691-4ca9-8735-a05690e6c2ec") // angular specific -> removed by default-filter
+                .header(org.springframework.http.HttpHeaders.CONTENT_TYPE, "application/hal+json")
                 .exchange()
                 .expectStatus().isEqualTo(HttpStatus.OK)
-                .expectHeader().valueMatches("Content-Type", "application/json")
-                .expectHeader().doesNotExist("WWW-Authenticate")
-                .expectHeader().valueMatches("Expires", "0")
+                .expectHeader().valueMatches(org.springframework.http.HttpHeaders.CONTENT_TYPE, "application/json")
+                .expectHeader().doesNotExist(org.springframework.http.HttpHeaders.WWW_AUTHENTICATE)
+                .expectHeader().valueMatches(org.springframework.http.HttpHeaders.EXPIRES, "0")
```
Comment (devtobi) [inline refarch-gateway/src/test/java/de/muenchen/refarch/gateway/route/BackendRouteTest.java:61]: Please don't use the FQCN here and instead import correctly. :)

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/route/BackendRouteTest.java [hunk 3]
```diff
@@ -48,20 +50,21 @@ void setup() {
...
                 .expectBody().jsonPath("$.testkey").isEqualTo("testvalue");
 
         verify(getRequestedFor(urlEqualTo("/remote/endpoint"))
-                .withoutHeader("Cookie")
-                .withoutHeader("X-SRF-TOKEN")
-                .withHeader("Content-Type", new EqualToPattern("application/hal+json")));
+                .withoutHeader(org.springframework.http.HttpHeaders.COOKIE)
+                .withoutHeader(XSRF_HEADER_NAME)
+                .withHeader(org.springframework.http.HttpHeaders.CONTENT_TYPE, new EqualToPattern("application/hal+json")));
```
Comment (devtobi) [inline refarch-gateway/src/test/java/de/muenchen/refarch/gateway/route/BackendRouteTest.java:67]: Please don't use the FQCN here and instead import correctly. :)

File: shared-files/pmd-ruleset.xml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    <rule ref="category/java/security.xml"/>
+    <rule ref="category/java/multithreading.xml">
+        <exclude name="UseConcurrentHashMap"/>
+    </rule>
+    <rule ref="category/java/bestpractices.xml/JUnitTestsShouldIncludeAssert">
+        <properties>
+            <!-- From Spring WebTestClient -->
+            <property name="extraAssertMethodNames" value="expectStatus,expectHeader" />
+        </properties>
+    </rule>
```
Comment (devtobi) [inline shared-files/pmd-ruleset.xml:39]: I discussed this with @DanielOber. Would it be possible to refactor the test cases to rely on assertions Methods directly known by PMD? (If it does not increase the LOC drastically).

**Review by devtobi — CHANGES_REQUESTED** [review]

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java [hunk 5]
```diff
@@ -30,51 +31,53 @@
...
                 .willReturn(aResponse()
                         .withStatus(HttpStatus.INTERNAL_SERVER_ERROR.value())
                         .withHeaders(new HttpHeaders(
-                                new HttpHeader("Content-Type", "application/json"),
-                                new HttpHeader("WWW-Authenticate", "Bearer realm=\"Access to the staging site\", charset=\"UTF-8\""),
-                                new HttpHeader("Expires", "Wed, 21 Oct 2099 07:28:06 GMT")))
+                                new HttpHeader(org.springframework.http.HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType()),
+                                new HttpHeader(org.springframework.http.HttpHeaders.WWW_AUTHENTICATE,
+                                        "Bearer realm=\"Access to the staging site\", charset=\"UTF-8\""),
+                                new HttpHeader(org.springframework.http.HttpHeaders.EXPIRES, "Wed, 21 Oct 2099 07:28:06 GMT")))
```
[the remark this answers is not in the document]
Comment (simonhir) [reply refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java:49]: As far as i know there is no alternative as `com.github.tomakehurst.wiremock.http.HttpHeaders` is already imported.

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java [hunk 6]
```diff
@@ -30,51 +31,53 @@
...
                         .withBody("{ \"testkey\" : \"testvalue\" }")));
 
         webTestClient.get().uri("/api/refarch-gateway-backend-service/remote").exchange()
                 .expectStatus().isEqualTo(HttpStatus.BAD_REQUEST)
-                .expectHeader().valueMatches("Content-Type", "application/json")
-                .expectHeader().doesNotExist("WWW-Authenticate")
-                .expectHeader().valueMatches("Expires", "0")
+                .expectHeader().valueMatches(org.springframework.http.HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType())
+                .expectHeader().doesNotExist(org.springframework.http.HttpHeaders.WWW_AUTHENTICATE)
+                .expectHeader().valueMatches(org.springframework.http.HttpHeaders.EXPIRES, "0")
```
[the remark this answers is not in the document]
Comment (simonhir) [reply refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java:56]: See above

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java [hunk 7]
```diff
@@ -30,51 +31,53 @@
...
                 .willReturn(aResponse()
                         .withStatus(HttpStatus.OK.value())
                         .withHeaders(new HttpHeaders(
-                                new HttpHeader("Content-Type", "application/json"),
-                                new HttpHeader("WWW-Authenticate", "Bearer realm=\"Access to the staging site\", charset=\"UTF-8\""),
-                                new HttpHeader("Expires", "Wed, 21 Oct 2099 07:28:06 GMT")))
+                                new HttpHeader(org.springframework.http.HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType()),
+                                new HttpHeader(org.springframework.http.HttpHeaders.WWW_AUTHENTICATE,
+                                        "Bearer realm=\"Access to the staging site\", charset=\"UTF-8\""),
+                                new HttpHeader(org.springframework.http.HttpHeaders.EXPIRES, "Wed, 21 Oct 2099 07:28:06 GMT")))
```
[the remark this answers is not in the document]
Comment (simonhir) [reply refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java:73]: See above

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java (hunk 1 above)
Comment (simonhir) [reply refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java:80]: See above

File: refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (devtobi) [reply refarch-gateway/src/test/java/de/muenchen/refarch/gateway/filter/GlobalBackend5xxTo400MapperTest.java:49]: Ah sorry. I did not see that. :) What a pity. ;)

**Review by devtobi — APPROVED** [review]
LGTM

**Pull request merged**